FirebaseAppCheck/Tests/Unit/AppAttestProvider/FIRAppAttestAPIServiceTests.m

```diff
@@ -341,6 +452,72 @@ - (id)URLValidationArgumentWithResource:(NSString *)resource {
+                                 response:(nullable GULURLSessionDataResponse *)response
+                                    error:(nullable NSError *)error
+                           andExpectToken:(nullable FIRAppCheckToken *)token {
+  id URLValidationArg = [self URLValidationArgumentWithResource:@"exchangeAppAttestAssertion"];
+
+  id bodyValidationArg = [OCMArg checkWithBlock:^BOOL(NSData *requestBody) {
+    NSDictionary<NSString *, id> *decodedData = [NSJSONSerialization JSONObjectWithData:requestBody
+                                                                                options:0
+                                                                                  error:nil];
+
+    XCTAssert([decodedData isKindOfClass:[NSDictionary class]]);
+
+    // Validate artifact field.
+    NSString *base64EncodedArtifact = decodedData[@"artifact"];
+    XCTAssert([base64EncodedArtifact isKindOfClass:[NSString class]]);
+
+    NSData *decodedAttestation = [[NSData alloc] initWithBase64EncodedString:base64EncodedArtifact
+                                                                     options:0];
+    XCTAssertEqualObjects(decodedAttestation, attestation);
+
+    // Validate challenge field.
+    NSString *base64EncodedChallenge = decodedData[@"challenge"];
+    XCTAssert([base64EncodedChallenge isKindOfClass:[NSString class]]);
+
+    NSData *decodedChallenge = [[NSData alloc] initWithBase64EncodedString:base64EncodedChallenge
+                                                                   options:0];
+    XCTAssertEqualObjects(decodedChallenge, challenge);
+
+    // Validate assertion field.
+    NSString *base64EncodedAssertion = decodedData[@"assertion"];
+    XCTAssert([base64EncodedAssertion isKindOfClass:[NSString class]]);
+
+    NSData *decodedAssertion = [[NSData alloc] initWithBase64EncodedString:base64EncodedAssertion
+                                                                   options:0];
+    XCTAssertEqualObjects(decodedAssertion, assertion);
+
+    return YES;
+  }];
+
+  FBLPromise *responsePromise = [FBLPromise pendingPromise];
+  if (error) {
+    [responsePromise reject:error];
+  } else {
+    [responsePromise fulfill:response];
+  }
+  OCMExpect([self.mockAPIService sendRequestWithURL:URLValidationArg
+                                         HTTPMethod:@"POST"
+                                               body:bodyValidationArg
+                                  additionalHeaders:@{@"Content-Type" : @"application/json"}])
+      .andReturn(responsePromise);
+
+  if (responsePromise.isFulfilled) {
+    FBLPromise *tokenPromise = [FBLPromise pendingPromise];
+    if (token) {
+      [tokenPromise fulfill:token];
+    } else {
+      NSError *tokenError = [NSError errorWithDomain:self.name code:0 userInfo:nil];
+      [tokenPromise reject:tokenError];
+    }
+    OCMExpect([self.mockAPIService appCheckTokenWithAPIResponse:response]).andReturn(tokenPromise);
+  }
```

Comment:
It feels like moving this part to a separate method and calling `expectTokenAPIRequestWithArtifact` and this new method one after another in the tests will result in more readable code. WDYT?

Reply:
Agreed... on it. I thought this part was challenging to write so thank you for the feedback.